cmd/experimental/kjobctl/pkg/builder/slurm_builder.go

```diff
@@ -0,0 +1,647 @@
+		memPerGPU,   // %[4]s
+		b.output,    // %[5]s
+		b.error,     // %[6]s
+		b.input,     // %[7]s
+		b.jobName,   // %[8]s
+		b.partition, // %[9]s
+	)
+}
+
+func (b *slurmBuilder) buildSlurmVariables() string {
+	nTasks := ptr.Deref(b.nTasks, 1)
+	nodes := ptr.Deref(b.nodes, 1)
+
+	return fmt.Sprintf(`export SLURM_ARRAY_JOB_ID=%[1]d       		# Job array’s master job ID number.
+export SLURM_ARRAY_TASK_COUNT=%[2]d  		# Total number of tasks in a job array.
+export SLURM_ARRAY_TASK_MAX=%[3]d    		# Job array’s maximum ID (index) number.
+export SLURM_ARRAY_TASK_MIN=%[4]d    		# Job array’s minimum ID (index) number.
+export SLURM_TASKS_PER_NODE=%[5]d    		# Number of tasks to be initiated on each node.
+export SLURM_CPUS_PER_TASK=%[6]s       		# Number of CPUs per task.
+export SLURM_CPUS_ON_NODE=%[7]s        		# Number of CPUs on the allocated node (actually pod).
+export SLURM_JOB_CPUS_PER_NODE=%[8]s   		# Count of processors available to the job on this node.
+export SLURM_CPUS_PER_GPU=%[9]s        		# Number of CPUs requested per allocated GPU.
+export SLURM_MEM_PER_CPU=%[10]s         	# Memory per CPU. Same as --mem-per-cpu .
+export SLURM_MEM_PER_GPU=%[11]s         	# Memory per GPU.
+export SLURM_MEM_PER_NODE=%[12]s        	# Memory per node. Same as --mem.
+export SLURM_GPUS=%[13]d                	# Number of GPUs requested (in total).
+export SLURM_NTASKS=%[14]d              	# Same as -n, –ntasks. The number of tasks.
+export SLURM_NTASKS_PER_NODE=%[15]d  		# Number of tasks requested per node.
+export SLURM_NPROCS=$SLURM_NTASKS       	# Same as -n, --ntasks. See $SLURM_NTASKS.
+export SLURM_NNODES=%[16]d            		# Total number of nodes (actually pods) in the job’s resource allocation.
+export SLURM_SUBMIT_DIR=%[17]s        		# The path of the job submission directory.
+export SLURM_SUBMIT_HOST=$HOSTNAME       	# The hostname of the node used for job submission.`,
+		1,                                  // %[1]d
+		b.arrayIndexes.Count(),             // %[2]d
+		b.arrayIndexes.Max(),               // %[3]d
+		b.arrayIndexes.Min(),               // %[4]d
+		nTasks,                             // %[5]d
+		getValueOrEmpty(b.cpusPerTask),     // %[6]s
+		getValueOrEmpty(b.cpusOnNode),      // %[7]s
+		getValueOrEmpty(b.cpusOnNode),      // %[8]s
+		getValueOrEmpty(b.cpusPerGpu),      // %[9]s
+		getValueOrEmpty(b.memPerCPU),       // %[10]s
+		getValueOrEmpty(b.memPerGPU),       // %[11]s
+		getValueOrEmpty(b.totalMemPerNode), // %[12]s
+		b.totalGpus,                        // %[13]s
+		nTasks,                             // %[14]d
+		nTasks,                             // %[15]d
+		nodes,                              // %[16]d
+		slurmPath,                          // %[17]s
+	)
+}
+
+func getValueOrEmpty(ptr *resource.Quantity) string {
+	if ptr != nil {
+		return ptr.String()
+	}
+
+	return ""
+}
+
+func (b *slurmBuilder) buildEntrypointCommand() string {
+	strBuilder := strings.Builder{}
+	strBuilder.WriteString("bash")
```

Comment:
does it have to be bash?
Couldn't the author decide to use another shell?

Reply:
This is a bash wrapper we use to initialize SLURM environment variables, similar to `srun` in SLURM. The wrapper will execute the user's bash script.

If I understood correct, user can use only SLURM `script.sh` file where he can execute any script that he want.

Reply:
IIUC, we are loading the script that the user passes into a volume. That's all good. My question is whether that script is necessarily written in bash.

Couldn't the user set `#! /usr/bin/zsh` or something else?

Reply:
Based on the [sbatch description](https://slurm.schedmd.com/sbatch.html#SECTION_DESCRIPTION), I'd say that the expected script by Slurm must be written in bash.

Reply:
> Couldn't the user set #! /usr/bin/zsh or something else?

We can run this script with chmod +x. And just execute script based on `#!/bin/bash`. On this case we can run not only bash script, but also any other scripts too (e.g. #!/usr/bin/env python3). 

Like this:
```
chmod +x /slurm/script
/slurm/script
```

@IrvingMg @alculquicondor WDYT?

Reply:
Good to know that slurm assumes bash, specifically.

But if it's not too hard to support any arbitrary scripting language, then I say let's do it.

Reply:
Done on 5a3f05c.